docs(docs-infra): Add query as an url query param to the api reference

mvdluit opened the pull request:
add an url query param to to perform a search on api references directly via the url

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Currently it's not possible to add a queryParam to https://angular.dev/api to link directly to a specific result set. This is possible on AIO. 

Issue Number: N/A


## What is the new behavior?
This brings the behavior inline with the old AIO API reference search. It's possible to directly link to a result set (e.g. https://angular.dev/api?query=inject). When searching via the input on the overview page, the search value will also be added to the queryParam.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mvdluit commented:
@JeanMeche As discussed on Discord. I did not yet add the functionality to also add the `type` queryParam when a `type` is selected. Please let me know if this is needed and I'd be happy to add it (it's basically a copy/paste of the `effect`).

JeanMeche commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -91,4 +101,11 @@ export default class ApiReferenceList {
+    let queryParams = {...this.route.snapshot.queryParams};
+    queryParams[key] = value;
+
+    this.router.navigate([], {queryParams});
```
We specifically didn't want that as it messes up the history.

mvdluit replied:
I changed it to use the Location service `replaceState`, hope this doesn;t have the same issue.

```ts
const newUrl = this.router.createUrlTree(['/api'], {queryParams: params}).toString();
this.location.replaceState(newUrl);
```

Also added the `type` queryParam when selected as that was the default behavior in AIO.

mvdluit commented:
@JeanMeche You mentioned the developer preview and experimental references yesterday. The current UI (input and toggle for `isDeprecated`) was created when the underlying data in `manifest.json` did not yet include properties for `isDeveloperPreview` and `isExperimental`. 

While re-thinking the UI is a bit out of scope for this PR, I do wonder if we shouldn't already include a `status` queryParam (as a temporary step before it is eventually fully included in the UI) that will show these references. It could be useful to be able to directly link from https://angular.dev/reference/releases#developer-preview to a list with current references in developer preview. This `status` queryParam should initially only be active when used in a direct link scenario (e.g. http://angular.dev/api?status=isdeveloperpreview) to avoid hiding items when people start manually searching. So we would have to reset it when a change to `query` or `type` takes place. Related issue: https://github.com/angular/angular/issues/55336

Just thinking out loud here, curious how you feel about this.

mvdluit commented:
Changed the PR to draft as I want to extract the `updateQueryParam` to a service, so it can be used in other places. (e.g. https://github.com/angular/angular/blob/fab673a1ddbca19ff9734f92a5ef0cc16be5708c/adev/src/app/features/update/update.component.ts#L186-L190)

atscott commented:
> Changed the PR to draft as I want to extract the updateQueryParam method to a service, so it can be used in other places. (e.g.

Do that in another follow-up refactoring PR if this is approved. Adding more to your current PR will make it more difficult to review and be approved. Using the `Location` service to hide browser URL updates from the Router should be used sparingly, if ever. I'd not approve the current change without first convincing myself that it's the right thing to do.

atscott commented:
Do you have examples of web pages with search that update the query as you're typing in the search input? I'm not convinced that this is a common or good user experience.

mvdluit commented:
> Do you have examples of web pages with search that update the query as you're typing in the search input? I'm not convinced that this is a common or good user experience.

It was the default behavior in the AIO API reference: https://v17.angular.io/api

atscott commented:
> It was the default behavior in the AIO API reference: https://v17.angular.io/api

Right, but that's not on its own a reason to do the same for the new docs site. Only if we decide it's a valuable feature that provides good UX.

mvdluit commented:
> > It was the default behavior in the AIO API reference: https://v17.angular.io/api
> 
> Right, but that's not on its own a reason to do the same for the new docs site. Only if we decide it's a valuable feature that provides good UX.

Would it be better if the param was only updated once the user was done typing? I would have to convert the signal to an observable and add a debounce maybe?

My use case involves the Angular Community Discord bot. It allows to search API references but it has a limit on how many results it can display. So when there are more than 10 results I want to direct users to https://angular.dev/api?query=
![api-command](https://github.com/user-attachments/assets/ff159ba8-38f8-451a-93cb-8e0250972756)

mvdluit commented:
@atscott This is actually not the first time this feature was discussed. You mentioned here that the url should be replaced: https://github.com/angular/angular/pull/52979#discussion_r1508021015 and in that PR there's no mention that this feature was not desirable (The actual feature was later removed from that PR due to issues getting it merged and is what lead me to create this PR)

atscott commented:
> @atscott This is actually not the first time this feature was discussed. You mentioned here that the url should be replaced: https://github.com/angular/angular/pull/52979#discussion_r1508021015 and in that PR there's no mention that this feature was not desirable (The actual feature was later removed from that PR due to issues getting it merged and is what lead me to create this PR)

Sure, but that doesn't at all mean I can't ask the question here.

mvdluit commented:
> > @atscott This is actually not the first time this feature was discussed. You mentioned here that the url should be replaced: [#52979 (comment)](https://github.com/angular/angular/pull/52979#discussion_r1508021015) and in that PR there's no mention that this feature was not desirable (The actual feature was later removed from that PR due to issues getting it merged and is what lead me to create this PR)
> 
> Sure, but that doesn't at all mean I can't ask the question here.

Didn't mean to imply that you couldn't/shouldn't ask. Just wanted to provide additional information.

mvdluit commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -91,4 +114,17 @@ export default class ApiReferenceList {
+    params[key] = value;
+    const newUrl = this.router.createUrlTree(['/api'], {queryParams: params}).toString();
+    this.location.replaceState(newUrl);
+  }
```
Changed it to use the Router. The scroll to top is disabled, but not sure if the implementation is how you expected it.

atscott commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -88,7 +96,52 @@ export default class ApiReferenceList {
+      {allowSignalWrites: true},
+    );
+
+    this.router.events
+      .pipe(
+        filter((e) => e instanceof NavigationEnd),
+        takeUntilDestroyed(),
+      )
+      .subscribe((_) => {
+        const info = this.router.lastSuccessfulNavigation?.extras.info as {[key: string]: boolean};
+        if (info) {
+          this.appScroller.disableScrolling = info['disableScrolling'];
+        }
+      });
+  }
```
Instead of having this here, you can add the check as another filter to the `AppScroller` implementation.

mvdluit replied:
I just did some additional testing as I wasn't sure about my initial implementation with the check inside the `tap`. I tried different things, but in the `tap` was the only way I could prevent scroll to top on the API reference page (when setting new params) while preserving normal scroll behavior when navigating to other pages (e.g. scroll down in the API reference overview page, enter a new query in the input and then navigate to `update guide`. If the check is added inside the subscribe, the `update guide` would not scroll to top on initial navigation like it should).

So this seems to behave like we want, but if there's a better way to implement the check I'd appreciate your guidance.

atscott commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -88,7 +96,52 @@ export default class ApiReferenceList {
   }
+
+  updateQueryParams(key: string, value: string): void {
+    const params: Params = {[key]: value};
```
I think this function should instead just take `queryParams: Params` and avoid the weird intermediate.

mvdluit replied:
This method has been removed, the `router.navigate()` is now called from the effect.

atscott commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -88,7 +96,52 @@ export default class ApiReferenceList {
+    this.router.navigate([], {
+      queryParams: params,
+      queryParamsHandling: 'merge',
+      replaceUrl: true,
```
The `fragment` should probably be retained as well.

mvdluit replied:
`preserveFragment: true,` added.

atscott commented on adev/src/app/features/references/api-reference-list/api-reference-list.component.ts:
```diff
@@ -88,7 +96,52 @@ export default class ApiReferenceList {
   itemTypes = Object.values(ApiItemType);
 
+  constructor() {
+    effect(
+      () => {
+        if (this.query() !== undefined) {
+          this.updateQueryParams('query', this.query() as string);
+        }
+      },
+      {allowSignalWrites: true},
+    );
+
+    effect(
+      () => {
+        if (this.type() !== undefined) {
+          this.updateQueryParams('type', this.type() as string);
+        }
+      },
+      {allowSignalWrites: true},
+    );
```
Should this just be 1 effect that updates both the query and the type? At that point, you don't even need the shared helper function. Also, shouldn't this remove the type or query params if they are `undefined`?

mvdluit replied:
Yes, I can combine this in 1 effect. Just to make sure, you mean that the `router.navigate()` to set new params can also happen directly in the effect right?

I'm not sure about removing the param in case of `undefined`. @JeanMeche Whats your opinion on this?

mvdluit replied:
Ok, I've made a change to the effects. `undefined` and `''` will now both result in the param being removed (overwritten with an empty `Params` object). Removed `queryParamsHandling: 'merge'` since both queryparams are set at the same time now. I'm not sure `undefined` can happen once the params are available, but this should prevent empty params in any case.

```ts
effect(
  () => {
    let params: Params = {};
    if (this.query() !== undefined && this.query() !== '') {
      params['query'] = this.query();
    }
    if (this.type() !== undefined && this.type() !== '') {
      params['type'] = this.type();
    }
    this.updateQueryParams(params);
  },
  {allowSignalWrites: true},
);
```
I'll look into the other things tomorrow as it's getting late for me.

atscott replied:
> you mean that the router.navigate() to set new params can also happen directly in the effect right

Yep, that's right.

> I'm not sure about removing the param in case of undefined.

If you don't delete it, then it shows up as empty in the URL.
![image](https://github.com/user-attachments/assets/f6281920-f3f3-42f1-9d63-06f5e9087cd5)

mvdluit replied:
Yup, empty params should no longer be possible with the change I posted above. Will push once I've addresses all the feedback.

atscott commented:
I think it might be a good idea to update the section links to preserve the query params. WDYT? 
https://github.com/angular/angular/blob/83117f1a0d4039f89f50a458e6c62ad3e2bee584/adev/src/app/features/references/api-items-section/api-items-section.component.html#L6

mvdluit commented:
> I think it might be a good idea to update the section links to preserve the query params. WDYT?
> 
> https://github.com/angular/angular/blob/83117f1a0d4039f89f50a458e6c62ad3e2bee584/adev/src/app/features/references/api-items-section/api-items-section.component.html#L6

Yup, think that is a good idea. I will make the change.

atscott reviewed (approved):
LGTM for the code. I think we have a clean and workable solution now. Should get approval from @twerske as well

mvdluit commented:
> LGTM for the code. I think we have a clean and workable solution now. Should get approval from @twerske as well

Yes, thank you for all the feedback and help Andrew!

I also want to mention a potential follow up PR to this one, which would add the ability to filter on status (similar to the status filtering on AIO). The current UI was created when `isDeveloperPreview` and `isExperimental` were not yet available in the manifest.json. I think this would add value, being able to link directly to a list of API references in developer preview or experimental. It would solve this issue https://github.com/angular/angular/issues/55336 by directly linking to the references from https://angular.dev/reference/releases#developer-preview and https://angular.dev/reference/releases#experimental

My current idea would be to replace the `deprecated` toggle with a button dropdown with the following options: `all (=default) | stable | deprecated | developer preview | experimental`. This would change the current behavior of initially hiding `deprecated` references and show all references on the overview page still with an indicator for the `deprecated` references.

Would like to hear opinions on this (also happy to take the discussion about this somewhere else if more appropriate).

cc. @JeanMeche

atscott commented:
> Would like to hear opinions on this (also happy to take the discussion about this somewhere else if more appropriate).

Could you open a new issue for this? That'd be the appropriate place to discuss feature requests

mvdluit commented:
Follow issue: https://github.com/angular/angular/issues/57119

JeanMeche commented:
Don't you think the "No API items found" text should go below instead of on the side ? 

<img width="872" alt="Screenshot 2024-07-31 at 00 21 01" src="https://github.com/user-attachments/assets/654d9a23-307c-4133-8dc3-3bf0fba94993">

mvdluit commented:
> Don't you think the "No API items found" text should go below instead of on the side ?
> 
> <img alt="Screenshot 2024-07-31 at 00 21 01" width="872" src="https://private-user-images.githubusercontent.com/1300985/353632953-654d9a23-307c-4133-8dc3-3bf0fba94993.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNzg0NDEsIm5iZiI6MTcyMjM3ODE0MSwicGF0aCI6Ii8xMzAwOTg1LzM1MzYzMjk1My02NTRkOWEyMy0zMDdjLTQxMzMtOGRjMy0zYmYwZmJhOTQ5OTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDczMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MzBUMjIyMjIxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGU3NWQyNzc4NGE2ZDIxODMwZTRiMjE4NmE5NGQzY2M0OWMzMWI1Yzk2M2YyOWVkNWQ3ODUwMWNhNDU4ZWUwMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nmlIQ9KYJtD3r-qCZ1pGt2roSAkCuO0aIi3pU5znUr8">

Sure, something like this? 
![apiref_noitems](https://github.com/user-attachments/assets/c0b2aa05-3df4-40ff-903c-929cb673fc9f)
`flex-basis: 100%` on `.adev-reference-list-empty`

Since it's not directly related to the query params, should this be it's own PR?

JeanMeche commented:
My bad you're right, I thought initially it was related to your changes.

twerske reviewed (approved):
I'm approving, I think for linking to docs this is an ok additional feature for the intended use case of helper bots on social platforms allowing users to see multiple docs pages. 

Though I'd suggest that linking to the specific docs page or reference is a better experience.

One suggested additional feature would actually be making the search UI have a url query param so that the linked page is the search widget and not the search API docs page.

alxhub commented:
This PR was merged into the repository by commit 564a8d587148f39dfdde58d5c6d5a4c921d116c4.

The changes were merged into the following branches: main, 18.2.x

JeanMeche commented:
We had to revert that change because it broke the API search (see #57669)

mvdluit commented:
@JeanMeche This can be solved by wrapping the initial input focus in an `afterRender` instead of `afterNextRender`. Do you see any issues with this?

JeanMeche commented:
We would need to probably change de behavior of this function https://github.com/angular/angular/blob/68e5370a66633f4b069d6adffa95c2ea94291820/adev/src/app/app.component.ts#L85

mvdluit commented:
> We would need to probably change de behavior of this function
> 
> https://github.com/angular/angular/blob/68e5370a66633f4b069d6adffa95c2ea94291820/adev/src/app/app.component.ts#L85

exclude the api route?

mvdluit commented:
@JeanMeche excluding the api route when there is at least one query param works, thoughts?

```ts
focusFirstHeading(): void {
  if (!this.isBrowser || this.router.url.includes('/api?')) {
    return;
  }

  const h1 = this.document.querySelector<HTMLHeadingElement>('h1');
  h1?.focus();
}
```

atscott commented:
Should the focus really be happening at all if there is a navigation where only the query params or fragment changes? We do something similar with the animation between routes, where it's skipped if the page isn't changing: https://github.com/angular/angular/blob/4fa25cf328a79292300c509084847dad459bfd41/adev/src/app/app.config.ts#L60-L71

mvdluit commented:
> Should the focus really be happening at all if there is a navigation where only the query params or fragment changes? We do something similar with the animation between routes, where it's skipped if the page isn't changing:
> 
> https://github.com/angular/angular/blob/4fa25cf328a79292300c509084847dad459bfd41/adev/src/app/app.config.ts#L60-L71

I think it makes sense to skip the focus for navigation where only fragments/queryParams change. I have it working like this now, but not sure if there's maybe a better way to get the `urlTree` for same page navigation.

```ts
focusFirstHeading(): void {
  if (!this.isBrowser) {
    return;
  }

  // ignore focus when only fragment/queryParams change
  const sameTree = createUrlTreeFromSnapshot(this.activatedRoute.snapshot, []);
  if (
    this.router.isActive(sameTree, {
      paths: 'exact',
      matrixParams: 'exact',
      fragment: 'ignored',
      queryParams: 'ignored',
    })
  ) {
    return;
  }

  const h1 = this.document.querySelector<HTMLHeadingElement>('h1');
  h1?.focus();
}
```

atscott commented:
It feels like what you have probably just always disables the focus though because it's being called on `NavigationEnd` when the route has already become active.

mvdluit committed:
docs(docs-infra): do not focus heading when only fragments or query params changes

setting heading focus when query params change, causes the api overview input to lose focus on every character entered. while staying on the same route, re-setting focus does not make sense

mvdluit commented:
Apologies for the mess with the rebase. I'm not sure what went wrong. Thank you @JeanMeche for helping out.

atscott commented on adev/src/app/app.component.ts:
```diff
@@ -108,12 +116,23 @@ export class AppComponent implements OnInit {
     this.router.events
       .pipe(
-        filter((e): e is NavigationEnd => e instanceof NavigationEnd),
+        filter((e): e is RoutesRecognized => e instanceof RoutesRecognized),
```
This isn't quite right. Guards and resolvers will delay the navigation and the focus will happen before the new route is rendered. On option might be a `switchMap` to a `NavigationEnd` after filtering on the condition that's in the subscription

atscott commented on adev/src/app/app.component.ts:
```diff
@@ -108,9 +116,24 @@ export class AppComponent implements OnInit {
+          return this.router.events.pipe(
+            filter((e): e is NavigationEnd => e instanceof NavigationEnd),
+            // Skip first emission, cause on the initial load we would like to `Skip to main content` popup when it's focused
+            skip(1),
```
This `skip(1)` doesn't belong here actually. It should probably be changed to a `skipUntil(router.events.filter(e => e instanceof NavigationEnd))` at the top of this pipe.

mvdluit replied:
I'm struggling to get it to work with `skipUntil`. Once I add it, the filter that checks same route navigation no longer works. Did I add it in the wrong place, or am I missing something else?

```ts
private focusFirstHeadingOnRouteChange(): void {
  this.router.events
    .pipe(
      filter((e): e is RoutesRecognized => e instanceof RoutesRecognized),
      filter((e) => {
        // do not set focus when only fragments or queryParams change
        const tree = createUrlTreeFromSnapshot(e.state.root, []);
        return this.router.isActive(tree, {
          paths: 'exact',
          matrixParams: 'exact',
          fragment: 'ignored',
          queryParams: 'ignored',
        });
      }),
      switchMap(() => {
        return this.router.events.pipe(
          skipUntil(
            this.router.events.pipe(
              filter((e): e is NavigationEnd => e instanceof NavigationEnd),
            ),
          ),
        );
      }),
    )
    .subscribe(() => {
      this.focusFirstHeading();
    });
  }
```

Also I saw https://github.com/angular/angular/pull/57695/commits/19a129683904db21ee7be152753e41eae1492170. I assume that is a temporary solution and this is still needed?

atscott replied:
> Did I add it in the wrong place

Yes, it needs to go at the very top, before `RoutesRecognized`,

> I assume that is a temporary solution and this is still needed?

Yes, it's temporary

mvdluit replied:
Apologies, i've been distracted with some other things. I've placed the `skipUntil` on top and it's working as expected now. Is there anything else that needs to be done to land this fix?

cc. @JeanMeche

atscott replied:
>  I've placed the skipUntil on top and it's working as expected now. Is there anything else that needs to be done to land this fix?

Yes, the filter was changed to look for `RoutesRecognized` but needs to be reverted back to `NavigationEnd`.

atscott replied:
Or rather, what the comment in the review says: after filtering, it needs to wait for `NavigationEnd` still so switchMapping to `NavigationEnd` should work.

mvdluit replied:
When I add a `switchMap`, `this.focusFirstHeading();` is always called. If the filter returns false (same route navigation) it correctly ignores the switchMap, but there's still an emission that triggers the method call in the `subscribe`. This does not happen when I remove the `switchMap` and I'm not sure what I'm doing wrong.

```ts
this.router.events
  .pipe(
    skipUntil(
      this.router.events.pipe(filter((e): e is NavigationEnd => e instanceof NavigationEnd)),
    ),
    filter((e): e is RoutesRecognized => e instanceof RoutesRecognized),
    filter((e) => {
      // do not set focus when only fragments or queryParams change
      const tree = createUrlTreeFromSnapshot(e.state.root, []);
      return !this.router.isActive(tree, {
        paths: 'exact',
        matrixParams: 'exact',
        fragment: 'ignored',
        queryParams: 'ignored',
      });
    }),
    switchMap(() => {
      return this.router.events.pipe(
        filter((e): e is NavigationEnd => e instanceof NavigationEnd),
      );
    }),
  )
  .subscribe(() => {
    this.focusFirstHeading();
  });
```

As an alternative you mentioned that `RoutesRecognized` should be reverted back to `NavigationEnd`, but I was only able to get the filter to work correctly with the `ActivatedRouteSnapshot` available on `RoutesRecognized.state.root`. Is there another snapshot besides `ActivatedRoute` that I could use?

e.g. This will never call the `this.focusFirstHeading()` method, even when there is a route change:
```ts
this.router.events
  .pipe(
    skipUntil(
      this.router.events.pipe(filter((e): e is NavigationEnd => e instanceof NavigationEnd)),
    ),
    filter((e): e is NavigationEnd => e instanceof NavigationEnd),
    filter(() => {
      // do not set focus when only fragments or queryParams change
      const tree = createUrlTreeFromSnapshot(this.activatedRoute.snapshot, []);
      return !this.router.isActive(tree, {
        paths: 'exact',
        matrixParams: 'exact',
        fragment: 'ignored',
        queryParams: 'ignored',
      });
    }),
  )
  .subscribe(() => {
    this.focusFirstHeading();
  });
```

JeanMeche commented:
This is already implemented. 
https://angular.dev/api?query=test
